### ISSUE TITLE
NVDA saves config even if Save configuration on exit is not checked

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#10288]
### Steps to reproduce:

Preliminary configuration: insure that NVDA does not save config when exiting:

* In general ssettings, uncheck "Save configuration on exit" check box and validate
* Save this new configuration: NVDA -> Save configuration

Steps to reproduce the issue:

* With NVDA+P, change your symbol level. E.g. if you usually are working with symbol level "som", press two times to get symbol level "all".
* Go to NVDA settings, General panel (NVDA+Ctrl+G)
* Change the language, e.g. if your UI is usually in English, select French
* Validate and select "Restart now" when you are prompted to restart NVDA

### Actual behavior:

NVDA language has switched to French, but symbol level is set to "all".

### Expected behavior:

NVDA language has switched to French, but symbol level remains to "some".

#### NVDA installed/portable/running from source:
NVDA installed

#### NVDA version:

2019.2

#### Windows version:

Win 10 1709

#### Name and version of other software in use when reproducing the issue:

N/A

#### Other information about your system:

N/A

#### Does the issue still occur after restarting your PC?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No

**Comment by JulienCochuyt** [comment]
The described behavior is indeed what is implemented for quite a long time now.
The configuration is saved altogether as soon as the user confirms restarting upon language change, which allows NVDA to take this new setting into account.

Alternatives might be:
 - To explicitly warn the user that the configuration will indeed be saved.
 - To detect the configuration is dirty (edited but not yet saved) and only warn then.
 - To offer the possibility to only save the language change, dropping all other modifications to the configuration.

@feerrenrut, what do you think?

**Comment by JulienCochuyt** [comment]
@CyrilleB79, this issue and #10289 really are two sides of the same coin.
I agree the current behavior is potentially questionable, but I guess we'll be more efficient in concentrating the discussion into a single thread.
As I feel the present issue deals with the least expected side, I propose renaming this one something like "NVDA forcibly saves configuration when restarting on language change" and closing #10289.

**Comment by CyrilleB79** [comment]
@JulienCochuyt, thanks.
I have closed #10289 to have all discussion in this ticket as recommended by you; that will indeed be easier to track.
So as explained in #10289, let's also take into account that changing language and clicking "Restart later" does not make so much sense unless you save manually your configuration afterwards.

**Comment by JulienCochuyt** [comment]
> So as explained in #10289, let's also take into account that changing language and clicking "Restart later" does not make so much sense unless you save manually your configuration afterwards.

I guess warning the user, when it chooses to restart later, that the configuration shall be saved for this change to take effect might be sufficient.

**Comment by CyrilleB79** [comment]
> I guess warning the user, when it chooses to restart later, that the configuration shall be saved for this change to take effect might be sufficient

Or maybe warn the user when he leaves NVDA if the language in the saved config (on disk) is not the same as the language currently selected in current config.

**Comment by JulienCochuyt** [comment]
> > I guess warning the user, when it chooses to restart later, that the configuration shall be saved for this change to take effect might be sufficient
> 
> Or maybe warn the user when he leaves NVDA if the language in the saved config (on disk) is not the same as the language currently selected in current config.

Bigger impact, but doable. I'm not fully convinced it is worth the effort, though.

**Comment by XLTechie** [comment]
I prefer the option that only saves the portion of config related to language.

In sudocode:

Before restart, as last action:
If save on exit is on:
update language.
Save configuration.
Restart.
Else if save on exit is off:
Load configuration from disk.
Update language.
Save configuration.
Restart.

**Comment by JulienCochuyt** [comment]
> @XLTechie wrote:
> I prefer the option that only saves the portion of config related to language.

I agree this would bring IMHO the most expected behavior when agreeing to "Restart now" and "Save configuration on exit" is off.

What about when choosing to "Restart later"?
I propose that when "Save configuration on exit" is off, the current prompt that states "NVDA must be restarted for the new language to take effect." be complemented by "If you choose to restart later, mind saving the configuration first, or this change will be lost."

**Comment by CyrilleB79** [comment]
> What about when choosing to "Restart later"?
> I propose that when "Save configuration on exit" is off, the current prompt that states "NVDA must be restarted for the new language to take effect." be complemented by "If you choose to restart later, mind saving the configuration first, or this change will be lost."

I agree also with this.

**Comment by JulienCochuyt** [comment]
@CyrilleB79, @XLTechie: When prompted to restart, if saveConfigurationOnExit is off, the user might still want to save the whole configuration.
Thus, I propose to add a third button (only if saveConfigurationOnExit is off) and the following alternative disclaimer:
- Restart now: Discard other changes made to the configuration and restart NVDA with the new language setting
- Save configuration and restart now: Save all changes made to the configuration and restart NVDA.
- Restart later: Mind that if you do not save the configuration, the new language setting will be lost.

What do you think?

**Comment by JulienCochuyt** [comment]
Also, note that the current message has already been recently updated for 2019.1 as of #6416
